cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/MessagingOperationStatePublisher.java

```diff
@@ -16,47 +16,205 @@
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.messaging.client.StoreRequestBuilder;
 import io.cdap.cdap.messaging.spi.MessagingService;
+import io.cdap.cdap.proto.Notification;
+import io.cdap.cdap.proto.id.NamespaceId;
 import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.id.TopicId;
 import io.cdap.cdap.proto.operation.OperationError;
 import io.cdap.cdap.proto.operation.OperationResource;
+import io.cdap.cdap.proto.operation.OperationRunStatus;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Provides capabilities to send operation lifecycle specific messages.
  */
 public class MessagingOperationStatePublisher implements OperationStatePublisher {
 
   private final MessagingService messagingService;
+  private final RetryStrategy retryStrategy;
+  private final List<TopicId> topicIds;
 
-  public MessagingOperationStatePublisher(MessagingService messagingService) {
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingOperationStatePublisher.class);
+
+  /**
+   * Create a publisher that writes to MessagingService topics depending on the message content.
+   *
+   * @param cConf configuration containing the topic prefix and number of partitions
+   * @param messagingService messaging service to write messages to
+   */
+  @Inject
+  public MessagingOperationStatePublisher(CConfiguration cConf, MessagingService messagingService) {
+    this(
+        messagingService,
+        cConf.get(Operation.STATUS_EVENT_TOPIC),
+        cConf.getInt(Operation.STATUS_EVENT_NUM_PARTITIONS),
+        RetryStrategies.fromConfiguration(
+            cConf, Operation.STATUS_RETRY_STRATEGY_PREFIX)
+    );
+  }
+
+  /**
+   * Create a publisher that writes to MessagingService topics depending on the message content.
+   *
+   * @param messagingService messaging service to write messages to
+   * @param topicPrefix prefix of the topic(s) to write to. If there is one topic, the prefix
+   *     will be the topic name. If there is more than one topic, the topic name will be the prefix
+   *     followed by the topic number
+   * @param numTopics number of topics to write to
+   * @param retryStrategy retry strategy to use for failures
+   */
+  @VisibleForTesting
+  public MessagingOperationStatePublisher(MessagingService messagingService,
+      String topicPrefix, int numTopics, RetryStrategy retryStrategy) {
     this.messagingService = messagingService;
+    this.topicIds =
+        numTopics == 1 ? Collections.singletonList(NamespaceId.SYSTEM.topic(topicPrefix)) :
```

Comment:
this logic is there in the program state publisher because we used to have a single topic before migrating to multiple topics. So we needed to be able to read from the pre-existing topic.

Over here, we don't have to worry a pre-existing topic, so it is better to always put the topic number at the end. This makes it straightforward if we ever want to increase the number of topics.